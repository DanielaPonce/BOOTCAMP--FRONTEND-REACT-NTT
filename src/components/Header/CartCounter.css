.cart-icon {
	cursor: pointer;
	position: relative;
	width: 30px;
	height: 30px;
}

.cart-icon .cart-count {
	position: absolute;
	top: -8px;
	right: -8px;
	background-color: var(--error-color);
	color: var(--secondary-color);
	border-radius: 50%;
	padding: 0.2rem 0.5rem;
	font-size: 0.8rem;
}

@media only screen and (max-width: 739px) {
	.cart-icon {
		position: absolute;
		top: 1.5rem;
		right: 2rem;
	}

}