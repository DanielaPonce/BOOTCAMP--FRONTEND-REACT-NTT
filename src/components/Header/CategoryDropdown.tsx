import { FC } from 'react';
import { Category } from '../../models/category.types';

type Props = {
	categories: Category[];
	onSelectCategory: (category: string) => void;
};

const CategoryDropdown: FC<Props> = ({ categories, onSelectCategory }) => {
	return (
		<select
			onChange={(e) => onSelectCategory(e.target.value)}
			defaultValue="all"
		>
			<option value="all">Todas las categorias</option>

			{categories.map(({ name, slug }) => (
				<option key={slug} value={slug}>
					{name}
				</option>
			))}
		</select>
	);
};

export default CategoryDropdown;
