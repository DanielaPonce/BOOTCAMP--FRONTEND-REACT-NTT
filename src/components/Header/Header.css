.header-container {
	background-color: var(--primary-color);
	position: relative;
	height: 12vh;
}

.header-container header {
	position: absolute;
	top: 5vh;
	width: 90vw;
	left: 50%; 

	transform: translate(-50%, 7%); 
	box-shadow: 5px 5px 15px rgba(0, 0, 0, 22%);
	background-color: var(--white);
	border-radius: 25px;
	color: var(--text-color);
	padding: 2em 24px;

	display: flex;
	justify-content: space-between;
	align-items: center;
}

header .search-container {
	display: flex;
	justify-content: center;
	align-items: center;
	gap: 8px;
}

.logo {
	display: flex;
	align-items: center;
	gap: 0.5rem;
	text-decoration: none;
	color: var(--text-color);
	cursor: pointer;
}

.logo h1:hover {
	color: var(--hover-button-color);
}


@media only screen and (min-width: 740px) and (max-width: 991px) {
	.logo h1 {
		font-size: 24px;
	}
	
}

@media only screen and (max-width: 739px) {
	.header-container header {
		flex-direction: column;
		top: 1vh;
		padding: 16px;
		gap: 16px;
		width: 85vw;
	}

	.logo h1 {
		margin: 0;
		font-size: 24px;
	}
}
