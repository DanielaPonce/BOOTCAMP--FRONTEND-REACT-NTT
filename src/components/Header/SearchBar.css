@media only screen and (max-width: 739px) {

	.header-container header .search-bar,
	.header-container header .search-bar input,
	.search-container select,
	.search-container {
		width: 100%;
		margin: 0;
	}

	.search-container {
		flex-direction: column;
	}
}
