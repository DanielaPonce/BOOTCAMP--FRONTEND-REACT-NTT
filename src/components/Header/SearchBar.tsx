import { FC } from 'react';
import { Constants } from '../../utils/constants';
import { StaticTexts } from '../../utils/staticTexts';

type Props = {
	onInputSearch: (search: string) => void;
};

const SearchBar: FC<Props> = ({ onInputSearch }) => {
	return (
		<div className="search-bar">
			<input
				type="search"
				placeholder={StaticTexts.searchBarPlaceholder}
				onInput={(e) =>
					onInputSearch((e.target as HTMLInputElement).value)
				}
			/>
		</div>
	);
};

export default SearchBar;
