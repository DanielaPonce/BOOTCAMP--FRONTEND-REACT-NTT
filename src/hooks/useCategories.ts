import { useEffect, useState } from 'react';
import { fetchAllCategories } from '../api/category.api';
import { Category } from '../models/category.types';

export const useCategories = () => {
	const [categories, setCategories] = useState<Category[]>([]);

	useEffect(() => {
		fetchAllCategories().then((data) => setCategories(data));
	}, []);

	return { categories };
};
