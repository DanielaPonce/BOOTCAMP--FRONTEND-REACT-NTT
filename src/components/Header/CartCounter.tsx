import { FC } from 'react';
import { useCart } from '../../hooks/useCart';
import { Link } from 'react-router';
import { RoutesConstants } from '../../utils/routes';

const CartCounter: FC = () => {
	const { itemCount } = useCart();

	return (
		<div className="cart-icon">
			<Link to={RoutesConstants.summary}>
				<img src="/src/assets/icons/cart-icon.svg" alt="cart icon" />
			</Link>

			{itemCount > 0 ? (
				<span id="cardCounter" className="cart-count">
					{itemCount}
				</span>
			) : (
				<></>
			)}
		</div>
	);
};

export default CartCounter;
