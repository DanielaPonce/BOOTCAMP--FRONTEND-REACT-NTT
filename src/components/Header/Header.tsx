import { FC } from 'react';
import { Link, useLocation } from 'react-router';
import { useCategories } from '../../hooks/useCategories';
import { StaticTexts } from '../../utils/staticTexts';
import CartCounter from './CartCounter';
import CategoryDropdown from './CategoryDropdown';
import SearchBar from './SearchBar';
import { RoutesConstants } from '../../utils/routes';
import { Images } from '../../utils/images';
import './Header.css';
import { useAuth } from '../../context/AuthContext';
import Button from '../Button/Button';

type HeaderProps = {
	onSearch?: (query: string) => void;
	onFilterCategory?: (category: string) => void;
};

const Header: FC<HeaderProps> = ({ onSearch, onFilterCategory }) => {
	const { user, logout } = useAuth();
	const location = useLocation();

	const { categories } = useCategories();

	const hideSearchAndDropdown = location.pathname === '/summary';

	return (
		<div className="header-container">
			<header>
				<div className="left-side">
					<Link to={RoutesConstants.Home} className="logo">
						<img
							src={Images.Logo}
							alt="Logo"
							width={45}
							height={45}
						/>
						<h1>{StaticTexts.AppTitle}</h1>
					</Link>
					{user && <span>Bienvenido: {user}</span>}
				</div>

				<div className="search-container">
					{hideSearchAndDropdown ? (
						<></>
					) : (
						<>
							<SearchBar onInputSearch={onSearch!} />
							<CategoryDropdown
								categories={categories}
								onSelectCategory={onFilterCategory!}
							/>
						</>
					)}
					<CartCounter />
					<Button
						type="button"
						className="logout-button"
						onClick={logout}
					>
						Cerrar Sesi√≥n
					</Button>
				</div>
			</header>
		</div>
	);
};

export default Header;
