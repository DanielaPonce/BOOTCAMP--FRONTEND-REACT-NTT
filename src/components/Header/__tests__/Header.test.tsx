import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router';
import { useCategories } from '../../../hooks/useCategories';
import { RoutesConstants } from '../../../utils/routes';
import Header from '../Header';
import { Category } from '../../../models/category.types';

jest.mock('../../../hooks/useCategories', () => ({
	__esModule: true,
	useCategories: jest.fn()
}));

jest.mock('../CartCounter', () => ({
	__esModule: true,
	default: jest.fn(() => <div data-testid="cart-counter">CartCounter</div>)
}));

jest.mock('../SearchBar', () => ({
	__esModule: true,
	default: jest.fn(({ onInputSearch }) => (
		<input
			data-testid="search-bar"
			type="search"
			onChange={(e) => onInputSearch(e.target.value)}
			placeholder="Search"
		/>
	))
}));

jest.mock('../CategoryDropdown', () => ({
	__esModule: true,
	default: jest.fn(({ categories, onSelectCategory }) => (
		<select
			data-testid="category-dropdown"
			onChange={(e) => onSelectCategory(e.target.value)}
		>
			{categories.map((category: Category) => (
				<option key={category.slug} value={category.slug}>
					{category.name}
				</option>
			))}
		</select>
	))
}));

describe('Header Component', () => {
	const mockCategories = [
		{ slug: 'category1', name: 'Category 1' },
		{ slug: 'category2', name: 'Category 2' }
	];

	beforeEach(() => {
		jest.clearAllMocks();
		(useCategories as jest.Mock).mockReturnValue({
			categories: mockCategories
		});
	});

	test('renders Header with logo, search bar, category dropdown, and cart counter', () => {
		render(
			<MemoryRouter>
				<Header />
			</MemoryRouter>
		);

		expect(screen.getByAltText('Logo')).toBeInTheDocument();

		expect(screen.getByTestId('search-bar')).toBeInTheDocument();

		expect(screen.getByTestId('category-dropdown')).toBeInTheDocument();

		expect(screen.getByTestId('cart-counter')).toBeInTheDocument();
	});

	test('hides search bar and dropdown on /summary route', () => {
		render(
			<MemoryRouter initialEntries={[RoutesConstants.Summary]}>
				<Header />
			</MemoryRouter>
		);

		expect(screen.queryByTestId('search-bar')).not.toBeInTheDocument();
		expect(
			screen.queryByTestId('category-dropdown')
		).not.toBeInTheDocument();

		expect(screen.getByTestId('cart-counter')).toBeInTheDocument();
	});

	test('calls onSearch when typing in the search bar', () => {
		const mockOnSearch = jest.fn();
		render(
			<MemoryRouter>
				<Header onSearch={mockOnSearch} />
			</MemoryRouter>
		);

		const searchBar = screen.getByTestId('search-bar');
		fireEvent.change(searchBar, { target: { value: 'query' } });

		expect(mockOnSearch).toHaveBeenCalledWith('query');
	});

	test('calls onFilterCategory when selecting a category', () => {
		const mockOnFilterCategory = jest.fn();
		render(
			<MemoryRouter>
				<Header onFilterCategory={mockOnFilterCategory} />
			</MemoryRouter>
		);

		const dropdown = screen.getByTestId('category-dropdown');
		fireEvent.change(dropdown, { target: { value: 'category1' } });

		expect(mockOnFilterCategory).toHaveBeenCalledWith('category1');
	});

	test('matches snapshot', () => {
		const { asFragment } = render(
			<MemoryRouter>
				<Header />
			</MemoryRouter>
		);
		expect(asFragment()).toMatchSnapshot();
	});
});
