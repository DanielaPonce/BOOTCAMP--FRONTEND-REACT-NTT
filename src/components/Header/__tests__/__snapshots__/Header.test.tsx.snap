// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="header-container"
  >
    <header>
      <a
        class="logo"
        data-discover="true"
        href="/"
      >
        <img
          alt="Logo"
          height="45"
          src="/src/assets/logo/logo.webp"
          width="45"
        />
        <h1>
          My Market
        </h1>
      </a>
      <div
        class="search-container"
      >
        <input
          data-testid="search-bar"
          placeholder="Search"
          type="search"
        />
        <select
          data-testid="category-dropdown"
        >
          <option
            value="category1"
          >
            Category 1
          </option>
          <option
            value="category2"
          >
            Category 2
          </option>
        </select>
        <div
          data-testid="cart-counter"
        >
          CartCounter
        </div>
      </div>
    </header>
  </div>
</DocumentFragment>
`;
